src/xwrap.rs

```diff
@@ -2,340 +2,248 @@
+                    (O::LSB_FIRST, 0xFF000000) | (O::MSB_FIRST, 0xFF) => 3,
                     _ => return None,
                 }
-            };
+            }};
         }
-        let red_offset = channel_offset!(img.red_mask);
-        let green_offset = channel_offset!(img.green_mask);
-        let blue_offset = channel_offset!(img.blue_mask);
-        let alpha_offset = channel_offset!(!(img.red_mask | img.green_mask | img.blue_mask));
-
-        // Wrap the pixel buffer into a slice
-        let size = (img.bytes_per_line * img.height) as usize;
-        let data = unsafe { slice::from_raw_parts(img.data as *const u8, size) };
+        let red_offset = channel_offset!(self.visual.red_mask);
+        let green_offset = channel_offset!(self.visual.green_mask);
+        let blue_offset = channel_offset!(self.visual.blue_mask);
+        let alpha_offset = channel_offset!(
+            !(self.visual.red_mask | self.visual.green_mask | self.visual.blue_mask)
+        );
 
         // Finally, generate the image object
-        Some(RgbaImage::from_fn(
-            img.width as u32,
-            img.height as u32,
-            |x, y| {
-                let offset = (y * img.bytes_per_line as u32 + x * bytes_per_pixel) as usize;
-                Rgba([
-                    data[offset + red_offset],
-                    data[offset + green_offset],
-                    data[offset + blue_offset],
-                    // Make the alpha channel fully opaque if none is provided
-                    if img.depth == 24 {
-                        0xFF
-                    } else {
-                        data[offset + alpha_offset]
-                    },
-                ])
-            },
-        ))
+        Some(RgbaImage::from_fn(self.w, self.h, |x, y| {
+            let offset = (y * bytes_per_line + x * bytes_per_pixel) as usize;
+            Rgba([
+                self.data[offset + red_offset],
+                self.data[offset + green_offset],
+                self.data[offset + blue_offset],
+                // Make the alpha channel fully opaque if none is provided
+                if self.format.depth == 24 {
+                    0xFF
+                } else {
+                    self.data[offset + alpha_offset]
+                },
+            ])
+        }))
     }
 
     fn to_image_buffer_rgb565(&self) -> Option<RgbaImage> {
-        let img = unsafe { &*self.handle };
-
-        if img.depth != 16 || img.bits_per_pixel != 16 {
+        if self.format.depth != 16 || self.format.bits_per_pixel != 16 {
             return None;
         }
         let bytes_per_pixel = 2;
 
-        // Wrap the pixel buffer into a slice
-        let size = (img.bytes_per_line * img.height) as usize;
-        let data = unsafe { slice::from_raw_parts(img.data as *const u8, size) };
+        let pad = match self.format.scanline_pad {
+            16 => 2,
+            32 => 4,
```

Comment:
I'd change these for `x @ 16 | 32 => x / 8` (may have gotten order wrong). Same as above. Makes it a bit clearer it's bits to bytes.